src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -115,34 +145,86 @@ internal async Task<IEnumerable<IProjectVersionedValue<RestoreData>>> RestoreAsy
-            // them to actually contains changes, only nominate if there are any.
-            byte[] hash = RestoreHasher.CalculateHash(restoreInfo);
+            ProjectRestoreInfo? restoreInfo = value.RestoreInfo;
+            bool success = false;
 
-            if (_latestHash != null && Enumerable.SequenceEqual(hash, _latestHash))
-                return true;
+            try
+            {
+                // Prevent extra restore under some conditions.
+                if (IsProjectConfigurationVersionOutOfDate(value.ConfiguredInputs))
+                {
+                    return false;
+                }
 
-            _latestHash = hash;
+                // Restore service always does work regardless of whether the value we pass 
+                // them to actually contains changes, only nominate if there are any.
+                byte[] hash = RestoreHasher.CalculateHash(restoreInfo!);
 
-            JoinableTask<bool> joinableTask = JoinableFactory.RunAsync(() =>
-            {
-                return NominateForRestoreAsync(restoreInfo, _projectAsynchronousTasksService.UnloadCancellationToken);
-            });
+                if (_latestHash != null && Enumerable.SequenceEqual(hash, _latestHash))
+                {
+                    SaveNominatedConfiguredVersions(value.ConfiguredInputs);
+                    return true;
+                }
+
+                _latestHash = hash;
+
+                _nugetRestoreStarted = true;
+                JoinableTask<bool> joinableTask = JoinableFactory.RunAsync(() =>
+                {
+                    return NominateForRestoreAsync(restoreInfo!, _projectAsynchronousTasksService.UnloadCancellationToken);
+                });
+
+                SaveNominatedConfiguredVersions(value.ConfiguredInputs);
+
+                _projectAsynchronousTasksService.RegisterAsyncTask(joinableTask,
+                                                                   ProjectCriticalOperation.Build | ProjectCriticalOperation.Unload | ProjectCriticalOperation.Rename,
+                                                                   registerFaultHandler: true);
 
-            _projectAsynchronousTasksService.RegisterAsyncTask(joinableTask,
-                                                               ProjectCriticalOperation.Build | ProjectCriticalOperation.Unload | ProjectCriticalOperation.Rename,
-                                                               registerFaultHandler: true);
+                // Prevent overlap until Restore completes
+                success = await joinableTask;
 
-            // Prevent overlap until Restore completes
-            bool success = await joinableTask;
+                lock (SyncObject)
+                {
+                    _nugetRestoreStarted = false;
+                }
 
-            HintProjectDependentFile(restoreInfo);
+                HintProjectDependentFile(restoreInfo!);
+            }
+            catch
```

Comment:
This is swallowing all exceptions from this, please remove as it will hide bugs, instead of faulting the block.

Reply:
we need delete the 'catch' block, only need a try/finally block here to reset the flag.

Reply:
catch block deleted in this [commit](https://github.com/dotnet/project-system/pull/7199/commits/d5a2a1f9b385e6a96d6d1c5fcbde5b7d0f19ebc5)